### PR TITLE
patch ipmikvm

### PR DESCRIPTION
**Pull request by cwrau**
Some things were not working out of the box, especially unpack200
Also move the APP_CACHE_DIR to a more standardised location, with fallback to the old path of course.
Add flock to the curl commands to make it possible to run this script in parallel.

~I didn't refactor/modernise everything, only the parts that had problems, but if you want I can also do that (shellcheck reports 27 problems)~

Now I just changed the above mentioned stuff

Created after https://github.com/ossobv/vcutil/pull/29, I don't know why that was closed 🤷

### CODE REVIEW

File: ipmikvm [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/usr/bin/env bash
```
[the remark this answers is not in the document]
Comment (cwrau) [reply ipmikvm:1]: shellcheck gave lots of errors because `In POSIX sh, 'local' is undefined.`, that's why I initially switched to using bash.

Also, newer tests, like `[[ -v VAR ]]`, does not exist in sh.

But if I should stay with `/bin/sh` I would remove the `local`s and convert my `[[`s to `test`

File: ipmikvm [hunk 2]
```diff
@@ -33,10 +33,29 @@
...
+    else
+        APP_CACHE_DIR="${XDG_CACHE_HOME:-${HOME?}/.cache}/ipmikvm"
+    fi
+fi
+
+export JAVA_HOME="${JAVA_HOME:-/usr/lib/jvm/java-8-openjdk}"
+if ! [[ -d "$JAVA_HOME" ]]; then
+    echo "$JAVA_HOME is missing" >/dev/stderr
+    exit 1
+fi
```
[the remark this answers is not in the document]
Comment (cwrau) [reply ipmikvm:49]: Oh, huh, I also have `/usr/bin/unpack200`, but that is a symlink to `/usr/lib/jvm/default-runtime/bin/unpack200` which is not necessarily java 8 and kinda shouldn't be, as java 8 is end of life 😅

I just assumed this was a relic of legacy times / setups where that still exists, but I can change that 👍

File: ipmikvm [hunk 3]
```diff
@@ -47,13 +66,13 @@ PASSES='ADMIN '
 # Alter this list with if you get more apparent ports when using ssh proxy:
 # http(80), https(443), ipmi(623), vnc(5900)
 PROXY_PORT_MAPPING='80:5980 443:5943 623:5923 5900:5959'
-PROXY_IP=  # dynamic?
+PROXY_IP= # dynamic?
```
[the remark this answers is not in the document]
Comment (cwrau) [reply ipmikvm:69]: > I have consistent formatting in this file, and you've just decided that your formatting is superior.

I didn't *decide* that, it's just that it gets auto-formatted 😅

If that bothers you I can remove that, I just thought these small changes are ok, as it's a lot of effort to turn off all the editor automations/formatting 😅

File: ipmikvm [hunk 4]
```diff
@@ -75,55 +94,61 @@ as_http() {
...
+    shift
     if test -n "$PROXY_IP"; then
         url=$(echo "$url" | sed -e '
-            s#^http://\([^/]*\)#http://'$PROXY_IP$WITH_HTTP_PORT'#
-            s#^https://\([^/]*\)#https://'$PROXY_IP$WITH_HTTPS_PORT'#')
+                s#^http://\([^/]*\)#http://'$PROXY_IP$WITH_HTTP_PORT'#
+                s#^https://\([^/]*\)#https://'$PROXY_IP$WITH_HTTPS_PORT'#')
     fi
-    command curl "$url" "$@"
+    flock -x "$APP_CACHE_DIR/lock" curl -fsSk "$url" "$@"
```
[the remark this answers is not in the document]
Comment (cwrau) [reply ipmikvm:104]: > This is strange. Why would you need locking on curl? Surely curl can be run multiple times. Does that mean that your target hosts cannot handle curls?
> 
> Do you have an explanation why you would need this?

If you run this script in parallel, the concurrent `curl`s calls overwrite each others' downloaded files 😅

I thought about some complex logic for locking, or moving the files, but this solution worked, wasn't complex and barely changed anything for single runs

> Also, you're adding `-k` here causing the caller to have no option but to silently ignore invalid certificates.

Yes, but *every single* `curl` call had the `-fsSk`, so I inlined that, but I can undo that if you want 👌

File: ipmikvm (hunk 3 above)
[the remark this answers is not in the document]
Comment (cwrau) [reply ipmikvm:69]: Ok, I've been trying to write the file without changing the formatting, but how?

Could you share your vim config? Even with the modeline `# vim: colorcolumn=80 tabstop=4 softtabstop=4 shiftwidth=4 expandtab:` stuff like the indented `printf`s or the `case`s still get indented differently 🤔

File: ipmikvm (hunk 4 above)
[the remark this answers is not in the document]
Comment (wdoekes) [reply ipmikvm]: You might be right that we're using -k everywhere, but we shouldn't have to. Ideally we have valid certs everywhere and only need -k when we don't. This is something on my to-fix list. But you couldn't know that ;)

[the remark this answers is not in the document]
Comment (wdoekes) [reply ipmikvm]: As for the flock. What files are they overwriting?

We have a `--cookie-jar=..` with a mktemp temp dir. So it shouldn't be that.

The curls themselves all output to stdout, so it shouldn't be that either.

Can you get to the bottom of that is being overwritten?

File: ipmikvm (hunk 3 above)
[the remark this answers is not in the document]
Comment (cwrau) [reply ipmikvm]: I just removed the modeline, I couldn't get it to work, it would always just format it differently.

I will just not format this file 😅

File: ipmikvm (hunk 2 above)
[the remark this answers is not in the document]
Comment (cwrau) [reply ipmikvm:55]: > I think I read somewhere that jars are zip compatible and could be unpacked without unpack200. But I hadn't bothered to look into that..

I tried it with `unzip`, but that failed, so I guess we stick with `unpack200`?

File: ipmikvm (hunk 4 above)
[the remark this answers is not in the document]
Comment (cwrau) [reply ipmikvm]: > As for the flock. What files are they overwriting?
> 
> We have a `--cookie-jar=..` with a mktemp temp dir. So it shouldn't be that.
> 
> The curls themselves all output to stdout, so it shouldn't be that either.
> 
> Can you get to the bottom of that is being overwritten?

The output files; `/home/cwr/.cache/ipmikvm/iKVM__V1.69.31.0x0/iKVM__V1.69.31.0x0.jar`

Multiple concurrent `curl [...] -o $file` to the same `$file` will get in each others' way

File: ipmikvm [hunk 5]
```diff
@@ -397,43 +426,45 @@ install_ikvm_application() {
...
+            echo "Unknown platform: $(uname -s)" >&2
+            exit 1
+            ;;
     esac
     local nativelibs=$(
-      echo "$launch_jnlp" | sed -e '
-        /<nativelib.*'$libprefix'/!d
-        s#.* href="\([^"]*\).jar".*version="\([^"]*\)".*#\1__V\2.jar#
-        s#.* href="\([^"]*\).jar".*#\1.jar#' | sort -u)
-    set -x
```
[the remark this answers is not in the document]
Comment (wdoekes) [reply ipmikvm]: Certainly. But in my experience that leaves a great big turd of info on the screen. And the relevant bits are here.

I don't necessarily mean debugging ipmikvm, but more debugging of why the downloading/running the jnlp/contents are not behaving as you want -- which unfortunately in my experience happens is required more often than you like.

File: ipmikvm (hunk 4 above)
[the remark this answers is not in the document]
Comment (wdoekes) [reply ipmikvm]: Funny. But in that case the unpack/gunzip could conflict as well.

I think you need locking on the entire `install_ikvm_application` instead. Locked on $app_cache_dir, which is where you do not want concurrent writes.

[the remark this answers is not in the document]
Comment (cwrau) [reply ipmikvm]: Done that, nice side-effect is, that the file only needs to be downloaded once, as every instance waits until a download is completed until it checks if it has to download the file 😁

One could further optimize this, but I think this is good enough and safe.

File: ipmikvm [hunk 6]
```diff
@@ -33,10 +33,38 @@
...
+    if ! command -v "$binary" >/dev/null; then
+        echo "'$binary' not found" >/dev/stderr
+        exit 1
+    fi
+done
+
+mkdir -p "$APP_CACHE_DIR"
+
+LOCKFD=4
+exec 4<>$APP_CACHE_DIR/lock
```
Comment (wdoekes) [inline ipmikvm:67]: This should be $LOCKFD.

File: ipmikvm [hunk 7]
```diff
@@ -319,10 +347,12 @@ get_app_jar() {
     umask 0077  # storing latest.jnlp here; might contain temp passwords
     test -d "$app_cache_dir" || mkdir -p "$app_cache_dir" || exit 1
     echo "$launch_jnlp" >"$app_cache_dir/latest.jnlp"
+    flock $LOCKFD
```
Comment (wdoekes) [inline ipmikvm:350]: Why not `exec $LOCKFD<>"$app_cache_dir/lock"` here?

**Review by wdoekes — COMMENTED** [review]
Almost there :)

Can you turn this into two or three commits? The APP_CACHE_DIR changes do not depend on the locks. And the gunzip -f depend on nothing either.

**Comment by wdoekes** [comment]
For the record, I tried shellcheck just now, and I see this:

```
In ipmikvm line 37:
test -z "$HOME" && echo 'missing $HOME' && exit 1
                        ^-------------^ SC2016 (info): Expressions don't expand in single quotes, use double quotes for that.
```
^- false positive.
```
In ipmikvm line 56:
    test ${1:-1} -ne 0 && exec >&2  # non-zero? write to stderr
         ^-----^ SC2086 (info): Double quote to prevent globbing and word splitting.
```
Okay.. this could be a thing. But the only caller of the function is in the same file. And it gets called with 0 or 1.
```
In ipmikvm line 67:
    echo "~/.config/ipmikvm/dict - as ALIAS ADDRESS USER PASS - one per"
          ^-- SC2088 (warning): Tilde does not expand in quotes. Use $HOME.
```
^- false positive
```
In ipmikvm line 70:
    exit ${1:-1}
         ^-----^ SC2086 (info): Double quote to prevent globbing and word splitting.
```
Same as previous.
```
In ipmikvm line 78:
    local url="$1"; shift
    ^-------^ SC3043 (warning): In POSIX sh, 'local' is undefined.
```
I've discussed about local being available.
```
In ipmikvm line 81:
            s#^http://\([^/]*\)#http://'$PROXY_IP$WITH_HTTP_PORT'#
                                        ^-------^ SC2086 (info): Double quote to prevent globbing and word splitting.
                                                 ^-------------^ SC2086 (info): Double quote to prevent globbing and word splitting.
```
This is possible. But again, I'm the one filling those arguments, so I know there is no space in there.

```
In ipmikvm line 137:
    read discard
    ^--^ SC2162 (info): read without -r will mangle backslashes.
         ^-----^ SC2034 (warning): discard appears unused. Verify use (or export if used externally).
```
Yes, but I'm discarding it. So I don't care.
```
In ipmikvm line 203:
            user=$(echo -n "$user" | base64 -w0 | sed -e 's/=/%3D/g;s/+/%2B/g')
                        ^-- SC3037 (warning): In POSIX sh, echo flags are undefined.
```
:+1: True. I should use printf for this.
```
In ipmikvm line 217:
            if test $ok_heur -gt 0 && test $fail_heur -gt 0; then
                    ^------^ SC2086 (info): Double quote to prevent globbing and word splitting.
                                           ^--------^ SC2086 (info): Double quote to prevent globbing and word splitting.
```
^- populated by me with `wc -l` or `grep -c`
```
In ipmikvm line 476:
    echo "(remote probably uses" $possible_ports "for ports)" >&2
                                 ^-------------^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean: 
    echo "(remote probably uses" "$possible_ports" "for ports)" >&2
```
Intentionally no "", with double quotes the echo will print linefeeds where I do not want any.
```
In ipmikvm line 469:
LF='
^-- SC2034 (warning): LF appears unused. Verify use (or export if used externally).
```
It is used, see `${LF}`.
```
In ipmikvm line 480:
    eval set -- $(get_arguments "$JNLP" | replace_arguments \
                ^-- SC2046 (warning): Quote this to prevent word splitting.
```
I'm pretty sure that adding double quotes here will break things too.

**In short, this shellcheck shit is just shit. Not using.**

File: ipmikvm (hunk 6 above)
Comment (cwrau) [reply ipmikvm:67]: That sadly doesn't work 😥;

```
./ipmikvm: line 67: exec: 4: not found
```

I _could_ use `eval exec "$LOCKFD"'<>"$APP_CACHE_DIR/lock"'` but I found that very weird and hard to understand, especially since the `exec` is so close to the `LOCKFD` variable

File: ipmikvm (hunk 7 above)
Comment (cwrau) [reply ipmikvm:350]: You mean to be more "lazy"?

Yeah, I could do that 👍

Comment (wdoekes) [reply ipmikvm:350]: Not only lazy. With `$app_cache_dir` instead of `$APP_CACHE_DIR` the lock is for the _specific_ app cache, not the total (root) app cache.

If we look here:
```sh
JAR=$(get_app_jar "$JNLP" "$APP_CACHE_DIR")
```
and:
```sh
get_app_jar() {
    local launch_jnlp="$1"
    local app_cache_dir="$2"
# ...
    *)
        app_cache_dir="$app_cache_dir/$jar_base"
# ...
```
We see that the "$app_cache_dir" is specific to the jar_base, which could be `iKVM__V1.69.31.0x0`. For that jar_base, we might want locking. It's fine for others to download simultaneously.

Also, minor nit:
`exec 4<>$APP_CACHE_DIR/lock` is lacking the double quotes:
```console
$ APP='a b v'

$ exec 4<>$APP
bash: $APP: ambiguous redirect

$ exec 4<>"$APP"
```

**Comment by cwrau** [comment]
> Almost there :)
> 
> Can you turn this into two or three commits? The APP_CACHE_DIR changes do not depend on the locks. And the gunzip -f depend on nothing either.

Done 😁👍

**Comment by wdoekes** [comment]
Thanks for the effort!

**Comment by wdoekes** [comment]
Tagging generally happens when I want a fix of a tool deployed on our fleet. nbdig has had several improvements and is ready for general use, so that will likely be soon.

**Comment by cwrau** [comment]
> Tagging generally happens when I want a fix of a tool deployed on our fleet. nbdig has had several improvements and is ready for general use, so that will likely be soon.

Awesome, I've created https://aur.archlinux.org/packages/ipmikvm, which is also capable of creating alle the other tools if they are requested,, so I'll release a new version if you create a new tag 👌

**Comment by wdoekes** [comment]
@cwrau : new tag is created
